processing/src/main/java/io/druid/query/lookup/LookupReferencesManager.java

```diff
@@ -52,333 +55,399 @@
+        LOG.error(ex, "Failed to close lookup [%s].", e.getKey());
+      }
     }
+
+    LOG.info("LookupReferencesManager is stopped.");
   }
 
-  /**
-   * @param lookups {@link Map<String, LookupExtractorFactory>} containing all the lookup as one batch.
-   *
-   * @throws IllegalStateException if the manager is closed or if {@link LookupExtractorFactory#start()} returns false
-   */
-  public void put(Map<String, LookupExtractorFactory> lookups)
+  public void add(String lookupName, LookupExtractorFactoryContainer lookupExtractorFactoryContainer)
   {
-    Map<String, LookupExtractorFactory> failedExtractorFactoryMap = new HashMap<>();
-    try {
-      startStopLock.readLock().lockInterruptibly();
-    }
-    catch (InterruptedException e) {
-      throw Throwables.propagate(e);
-    }
-    try {
-      assertStarted();
-      for (Map.Entry<String, LookupExtractorFactory> entry : lookups.entrySet()) {
-        final String lookupName = entry.getKey();
-        final LookupExtractorFactory lookupExtractorFactory = entry.getValue();
-        if (lookupMap.containsKey(lookupName)) {
-          // Fail early without bothering to start
-          LOGGER.warn("lookup [%s] is not add, another lookup with the same name already exist", lookupName);
-          continue;
-        }
-        if (!lookupExtractorFactory.start()) {
-          failedExtractorFactoryMap.put(lookupName, lookupExtractorFactory);
-          continue;
-        }
-        if (null != lookupMap.putIfAbsent(lookupName, lookupExtractorFactory)) {
-          // handle race
-          LOGGER.warn("lookup [%s] is not add, another lookup with the same name already exist", lookupName);
-          if (!lookupExtractorFactory.close()) {
-            LOGGER.error("Failed to properly close stale lookup [%s]", lookupExtractorFactory);
+    Preconditions.checkState(lifecycleLock.awaitStarted(1, TimeUnit.MILLISECONDS));
+    addNotice(new LoadNotice(lookupName, lookupExtractorFactoryContainer));
+  }
+
+  public void remove(String lookupName)
+  {
+    Preconditions.checkState(lifecycleLock.awaitStarted(1, TimeUnit.MILLISECONDS));
+    addNotice(new DropNotice(lookupName));
+  }
+
+  private void addNotice(Notice notice)
+  {
+    atomicallyUpdateStateRef(
+        oldState -> {
+          if (oldState.pendingNotices.size() > 10000) { //don't let pendingNotices grow indefinitely
+            throw new ISE("There are too many [%d] pendingNotices.", oldState.pendingNotices.size());
           }
-          continue;
-        }
-        if (lookupSnapshotTaker != null) {
-          lookupSnapshotTaker.takeSnapshot(getAllAsList());
+
+          ImmutableList.Builder<Notice> builder = ImmutableList.builder();
+          builder.addAll(oldState.pendingNotices);
+          builder.add(notice);
+
+          return new LookupUpdateState(
+              oldState.lookupMap, builder.build(), oldState.noticesBeingHandled
+
+          );
         }
+    );
+    LockSupport.unpark(mainThread);
```

Comment:
If a long going update is firing in mainThread, it won't be parked and I don't think this will reach it. That can cause a missed notice.

Can a `BlockingQueue` be used instead?